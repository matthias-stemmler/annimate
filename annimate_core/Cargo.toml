[package]
name = "annimate_core"
version = "0.1.0"
edition.workspace = true

[build-dependencies]
cargo_metadata = "0.18"

[dependencies]
csv = "1.0"
graphannis = "3.2.1"
graphannis-core = "3.2.1"
itertools = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_with = "3.7"
thiserror = { workspace = true }
toml = "0.8"
urlencoding = "2.1"

[dev-dependencies]
indoc = "2.0"
insta = { version = "1.34", features = ["serde"] }
