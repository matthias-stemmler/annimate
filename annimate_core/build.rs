//! Build script for `annimate_core`.

use cargo_metadata::MetadataCommand;

fn main() {
    inject_graphannis_version();
}

fn inject_graphannis_version() {
    let metadata = MetadataCommand::new()
        .exec()
        .expect("Failed to run `cargo metadata`");

    let graphannis_version = metadata
        .packages
        .into_iter()
        .find(|p| p.name == "graphannis")
        .expect("`graphannis` package not found in Cargo metadata`")
        .version
        .to_string();

    println!("cargo::rustc-env=GRAPHANNIS_VERSION={graphannis_version}");
}
