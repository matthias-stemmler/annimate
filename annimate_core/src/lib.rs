use std::io::Write;
use std::path::{Path, PathBuf};

use anno::AnnoKeys;
use corpus::CorpusRef;
use format::export;
use graphannis::corpusstorage::CacheStrategy;
use import::{FilesystemEntity, ImportFormat, ImportableCorpus};
use itertools::Itertools;
use metadata::{CorpusSet, MetadataStorage};
use query::Query;
use serde::Serialize;

mod anno;
mod aql;
mod corpus;
mod error;
mod format;
mod import;
mod metadata;
mod node_name;
mod query;
mod util;
mod version;

pub use anno::{ExportableAnnoKey, ExportableAnnoKeys};
pub use aql::{QueryAnalysisResult, QueryNode, QueryNodes};
pub use error::AnnisExportError;
pub use format::{CsvExportColumn, CsvExportConfig, ExportFormat};
pub use graphannis::corpusstorage::CorpusInfo;
pub use graphannis::graph::AnnoKey;
pub use import::ImportedCorpus;
pub use query::{ExportData, ExportDataAnno, ExportDataText, QueryLanguage};
pub use version::{VersionInfo, VERSION_INFO};

pub struct Storage {
    corpus_storage: graphannis::CorpusStorage,
    metadata_storage: MetadataStorage,
}

impl Storage {
    pub fn from_db_dir<P>(db_dir: P) -> Result<Self, AnnisExportError>
    where
        P: AsRef<Path>,
    {
        let corpus_storage = graphannis::CorpusStorage::with_cache_strategy(
            db_dir.as_ref(),
            CacheStrategy::PercentOfFreeMemory(25.0),
            true,
        )?;

        let metadata_storage = MetadataStorage::from_db_dir(
            &db_dir,
            &corpus_storage
                .list()?
                .into_iter()
                .map(|c| c.name)
                .collect_vec(),
        )?;

        Ok(Self {
            corpus_storage,
            metadata_storage,
        })
    }

    pub fn corpora(&self) -> Result<Corpora, AnnisExportError> {
        let sets = self.metadata_storage.corpus_sets();

        let corpora = self
            .corpus_storage
            .list()?
            .into_iter()
            .sorted_by(|a, b| a.name.cmp(&b.name))
            .map(|c| {
                let included_in_sets = sets
                    .iter()
                    .filter(|(_, CorpusSet { corpus_names })| corpus_names.contains(&c.name))
                    .map(|(name, _)| name.clone())
                    .collect();

                Corpus {
                    name: c.name,
                    included_in_sets,
                }
            })
            .collect();

        Ok(Corpora {
            sets: sets.keys().cloned().collect(),
            corpora,
        })
    }

    pub fn delete_corpus(&self, corpus_name: &str) -> Result<(), AnnisExportError> {
        self.corpus_storage.delete(corpus_name)?;
        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            for CorpusSet { corpus_names } in corpus_sets.values_mut() {
                corpus_names.remove(corpus_name);
            }
        })?;
        Ok(())
    }

    pub fn import_corpora<F>(
        &self,
        paths: Vec<PathBuf>,
        on_status: F,
    ) -> Result<Vec<String>, AnnisExportError>
    where
        F: Fn(ImportStatusEvent),
    {
        let mut imported_corpus_names = Vec::new();

        let importable_corpora = import::find_importable_corpora(paths, |message| {
            on_status(ImportStatusEvent::Message {
                index: None,
                message: message.into(),
            });
        })?;

        on_status(ImportStatusEvent::CorporaFound {
            corpora: importable_corpora.iter().map_into().collect(),
        });

        for (index, corpus) in importable_corpora.into_iter().enumerate() {
            on_status(ImportStatusEvent::CorpusImportStarted { index });

            let result = import::import_corpus(&self.corpus_storage, corpus, |message| {
                on_status(ImportStatusEvent::Message {
                    index: Some(index),
                    message: message.into(),
                });
            });

            match result {
                Ok(imported_corpus) => {
                    imported_corpus_names.push(imported_corpus.imported_name.clone());

                    on_status(ImportStatusEvent::CorpusImportFinished {
                        index,
                        result: ImportCorpusResult::Imported {
                            corpus: imported_corpus,
                        },
                    });
                }
                Err(err) => {
                    on_status(ImportStatusEvent::CorpusImportFinished {
                        index,
                        result: ImportCorpusResult::Failed {
                            message: err.to_string(),
                        },
                    });
                }
            }
        }

        Ok(imported_corpus_names)
    }

    pub fn add_corpora_to_set<S>(
        &self,
        corpus_set_name: &str,
        corpus_names: &[S],
    ) -> Result<(), AnnisExportError>
    where
        S: AsRef<str>,
    {
        if corpus_names.is_empty() {
            return Ok(());
        }

        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            corpus_sets
                .entry(corpus_set_name.into())
                .or_default()
                .corpus_names
                .extend(corpus_names.iter().map(|c| c.as_ref().into()));
        })?;

        Ok(())
    }

    pub fn toggle_corpus_in_set(
        &self,
        corpus_set_name: &str,
        corpus_name: &str,
    ) -> Result<(), AnnisExportError> {
        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            if let Some(CorpusSet { corpus_names }) = corpus_sets.get_mut(corpus_set_name) {
                if !corpus_names.remove(corpus_name) {
                    corpus_names.insert(corpus_name.into());
                }
            }
        })?;

        Ok(())
    }

    pub fn validate_query<S>(
        &self,
        corpus_names: &[S],
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryAnalysisResult<()>, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(aql::validate_query(
            self.corpus_ref(corpus_names),
            aql_query,
            query_language,
        )?)
    }

    pub fn query_nodes(
        &self,
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryAnalysisResult<QueryNodes>, AnnisExportError> {
        Ok(aql::query_nodes(
            &self.corpus_storage,
            aql_query,
            query_language,
        )?)
    }

    pub fn exportable_anno_keys<S>(
        &self,
        corpus_names: &[S],
    ) -> Result<ExportableAnnoKeys, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(AnnoKeys::new(self.corpus_ref(corpus_names)).map(AnnoKeys::into_exportable)?)
    }

    pub fn segmentations<S>(&self, corpus_names: &[S]) -> Result<Vec<String>, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(anno::segmentations(self.corpus_ref(corpus_names))?)
    }

    pub fn export_matches<F, S, W>(
        &self,
        corpus_names: &[S],
        aql_query: &str,
        query_language: QueryLanguage,
        format: ExportFormat,
        mut out: W,
        on_status: F,
    ) -> Result<(), AnnisExportError>
    where
        F: Fn(ExportStatusEvent),
        S: AsRef<str>,
        W: Write,
    {
        let anno_keys = AnnoKeys::new(self.corpus_ref(corpus_names))?;
        let query = Query::new(self.corpus_ref(corpus_names), aql_query, query_language)?;
        let matches = query.find(format.get_export_data().cloned())?;

        on_status(ExportStatusEvent::Found {
            count: matches.total_count(),
        });

        export(
            format,
            matches,
            query.nodes(),
            anno_keys.format(),
            &mut out,
            |progress| on_status(ExportStatusEvent::Exported { progress }),
        )?;

        out.flush()?;

        Ok(())
    }

    fn corpus_ref<'a, S>(&'a self, corpus_names: &'a [S]) -> CorpusRef<'a, S> {
        CorpusRef::new(&self.corpus_storage, corpus_names)
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Corpora {
    corpora: Vec<Corpus>,
    sets: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Corpus {
    name: String,
    included_in_sets: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ExportStatusEvent {
    Found { count: usize },
    Exported { progress: f32 },
}

#[derive(Clone, Debug, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ImportStatusEvent {
    CorporaFound {
        corpora: Vec<ImportCorpus>,
    },
    CorpusImportStarted {
        index: usize,
    },
    CorpusImportFinished {
        index: usize,
        result: ImportCorpusResult,
    },
    Message {
        index: Option<usize>,
        message: String,
    },
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportCorpus {
    file_name: String,
    format: ImportFormat,
    trace: Vec<FilesystemEntity<String>>,
}

impl From<&ImportableCorpus> for ImportCorpus {
    fn from(importable_corpus: &ImportableCorpus) -> ImportCorpus {
        ImportCorpus {
            file_name: importable_corpus.file_name.clone(),
            format: importable_corpus.format,
            trace: importable_corpus.trace.clone(),
        }
    }
}

#[derive(Clone, Debug, Serialize)]
#[serde(tag = "type")]
#[serde(rename_all = "snake_case")]
#[serde(rename_all_fields = "camelCase")]
pub enum ImportCorpusResult {
    Imported { corpus: ImportedCorpus },
    Failed { message: String },
}
