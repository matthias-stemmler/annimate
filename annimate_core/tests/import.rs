use std::cell::{Cell, RefCell};
use std::fs::{self, File};
use std::io;
use std::path::{Path, PathBuf};

use annimate_core::{AnnimateError, ImportStatusEvent, Storage};
use zip::write::SimpleFileOptions;
use zip::{CompressionMethod, ZipArchive, ZipWriter};

const DATA_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/data");
const DB_DIR: &str = concat!(env!("CARGO_TARGET_TMPDIR"), "/tests/import/db");
const INPUT_DIR: &str = concat!(env!("CARGO_TARGET_TMPDIR"), "/tests/import/input");

#[test]
fn import() {
    let db_dir = Path::new(DB_DIR).join("import");

    let _ = fs::remove_dir_all(&db_dir);
    let storage = Storage::from_db_dir(db_dir).unwrap();

    let events = RefCell::new(Vec::new());

    storage
        .import_corpora(
            prepare_input_files(INPUT_DIR),
            |event| match event {
                ImportStatusEvent::Message { .. } => (),
                event => events.borrow_mut().push(event.clone()),
            },
            || false,
        )
        .unwrap();

    let corpora = storage.corpora().unwrap();

    insta::with_settings!(
        {
             omit_expression => true,
        },
        {
            insta::assert_yaml_snapshot!(events.into_inner(), {
                ".**.path" => "[path]"
            });
            insta::assert_debug_snapshot!(corpora);
        }
    );
}

#[rustfmt::skip]
// Creates the following folder structure:
//
// <root_path>
// ├── README.md                             # Non-corpus file
// ├── subtok.demo_relANNIS/                 # Corpus 1 subtok.demo (RelANNIS)
// └── test_dir/                             # Directory
//    ├── subtok.demo2_relANNIS/             # Corpus 2 subtok.demo2 (RelANNIS)
//    └── test_zip.zip                       # ZIP file
//        ├── empty.graphml                  # Corpus 3 empty (GraphML)
//        └── test_dir_in_zip/               # Directory in ZIP file
//            └── subtok.demo3_relANNIS.zip  # ZIP file in ZIP file
//                └── subtok.demo3_relANNIS/ # Corpus 4 subtok.demo3 (RelANNIS)
fn prepare_input_files<P>(root_path: P) -> Vec<PathBuf>
where
    P: AsRef<Path>,
{
    let root_path = root_path.as_ref();
    
    let _ = fs::remove_dir_all(root_path);
    fs::create_dir_all(root_path).unwrap();

    fs::copy(
        Path::new(DATA_DIR).join("README.md"),
        root_path.join("README.md"),
    )
    .unwrap();

    extract_zip(
        Path::new(DATA_DIR).join("subtok.demo_relANNIS.zip"),
        root_path,
    );

    fs::create_dir(root_path.join("test_dir")).unwrap();

    extract_zip(
        Path::new(DATA_DIR).join("subtok.demo2_relANNIS.zip"),
        root_path.join("test_dir"),
    );

    let mut zip_writer =
        ZipWriter::new(File::create(root_path.join("test_dir/test_zip.zip")).unwrap());

    zip_writer
        .start_file(
            "empty.graphml",
            SimpleFileOptions::default().compression_method(CompressionMethod::Stored),
        )
        .unwrap();

    let tempdir_graphml = tempfile::tempdir().unwrap();

    extract_zip(
        Path::new(DATA_DIR).join("empty_graphml.zip"),
        &tempdir_graphml,
    );

    io::copy(
        &mut File::open(tempdir_graphml.path().join("empty.graphml")).unwrap(),
        &mut zip_writer,
    )
    .unwrap();

    zip_writer
        .start_file(
            "test_dir_in_zip/subtok.demo3_relANNIS.zip",
            SimpleFileOptions::default().compression_method(CompressionMethod::Stored),
        )
        .unwrap();

    io::copy(
        &mut File::open(Path::new(DATA_DIR).join("subtok.demo3_relANNIS.zip")).unwrap(),
        &mut zip_writer,
    )
    .unwrap();

    vec![
        root_path.join("subtok.demo_relANNIS"),
        root_path.join("test_dir"),
        root_path.join("test_dir/subtok.demo2_relANNIS"), // testing deduplication
    ]
}

fn extract_zip<P, Q>(zip_path: P, output_dir: Q)
where
    P: AsRef<Path>,
    Q: AsRef<Path>,
{
    let mut archive = ZipArchive::new(File::open(zip_path).unwrap()).unwrap();
    archive.extract(output_dir).unwrap();
}

#[test]
fn import_cancelled_before_corpora_collected() {
    let db_dir = Path::new(DB_DIR).join("import_cancelled_before_corpora_collected");

    let _ = fs::remove_dir_all(&db_dir);
    let storage = Storage::from_db_dir(db_dir).unwrap();

    let result = storage.import_corpora(
        vec![
            Path::new(DATA_DIR).join("subtok.demo_relANNIS.zip"),
            Path::new(DATA_DIR).join("subtok.demo2_relANNIS.zip"),
        ],
        |_| (),
        || true,
    );

    assert!(matches!(result, Err(AnnimateError::Cancelled)));
    assert_eq!(storage.corpora().unwrap().corpus_count(), 0);
}

#[test]
fn import_cancelled_after_corpora_collected() {
    let db_dir = Path::new(DB_DIR).join("import_cancelled_after_corpora_collected");

    let _ = fs::remove_dir_all(&db_dir);
    let storage = Storage::from_db_dir(db_dir).unwrap();

    let cancel_requested = Cell::new(false);

    let imported_corpora = storage
        .import_corpora(
            vec![
                Path::new(DATA_DIR).join("subtok.demo_relANNIS.zip"),
                Path::new(DATA_DIR).join("subtok.demo2_relANNIS.zip"),
            ],
            |event| {
                if let ImportStatusEvent::CorporaFound { .. } = event {
                    cancel_requested.set(true);
                }
            },
            || cancel_requested.get(),
        )
        .unwrap();

    assert!(imported_corpora.is_empty());
    assert_eq!(storage.corpora().unwrap().corpus_count(), 0);
}

#[test]
fn import_cancelled_after_corpus_import_started() {
    let db_dir = Path::new(DB_DIR).join("import_cancelled_after_corpus_import_started");

    let _ = fs::remove_dir_all(&db_dir);
    let storage = Storage::from_db_dir(db_dir).unwrap();

    let cancel_requested = Cell::new(false);

    let imported_corpora = storage
        .import_corpora(
            vec![
                Path::new(DATA_DIR).join("subtok.demo_relANNIS.zip"),
                Path::new(DATA_DIR).join("subtok.demo2_relANNIS.zip"),
            ],
            |event| {
                if let ImportStatusEvent::CorpusImportStarted { .. } = event {
                    cancel_requested.set(true);
                }
            },
            || cancel_requested.get(),
        )
        .unwrap();

    assert_eq!(imported_corpora, ["subtok.demo"]);
    assert_eq!(storage.corpora().unwrap().corpus_count(), 1);
}
