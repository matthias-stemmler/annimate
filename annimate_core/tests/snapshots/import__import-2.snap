---
source: annimate_core/tests/import.rs
---
Corpora {
    corpora: [
        Corpus {
            name: "empty",
            included_in_sets: [],
        },
        Corpus {
            name: "subtok.demo",
            included_in_sets: [],
        },
        Corpus {
            name: "subtok.demo2",
            included_in_sets: [],
        },
        Corpus {
            name: "subtok.demo3",
            included_in_sets: [],
        },
    ],
    sets: [],
}
